torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -377,6 +378,318 @@ bool IsBlockReturnTypeSame(Node* n) {
+  double shape_ratio = 1.0;
+  for (auto i = 0; i < input_shape_size; i++) {
+    if (i < minus_one_pos) {
+      if (reshape[i] != 0) {
+        shape_ratio *= static_cast<double>(input_shape[i]) / reshape[i];
+      }
+    } else if (minus_one_pos > -1) {
+      if (i < input_shape_size - reshape_size + minus_one_pos + 1) {
+        shape_ratio *= static_cast<double>(input_shape[i]);
+      } else {
+        auto reshape_idx = i - input_shape_size + reshape_size;
+        if (reshape[reshape_idx] != 0) {
+          shape_ratio *=
+              static_cast<double>(input_shape[i]) / reshape[reshape_idx];
+        }
+      }
+    }
+  }
+
+  for (auto i = 0; i < minus_one_pos; i++) {
+    int64_t cur_shape = reshape[i] == 0 ? input_shape[i] : reshape[i];
+    final_shape.push_back(cur_shape);
+  }
+  final_shape.push_back(static_cast<int64_t>(std::round(shape_ratio)));
+  for (auto i = minus_one_pos + 1; i < reshape_size; i++) {
+    int64_t cur_shape = reshape[i] == 0
+        ? input_shape[i + input_shape_size - reshape_size]
+        : reshape[i];
+    final_shape.push_back(cur_shape);
+  }
+  return final_shape;
+}
+
+void UpdateRank(Value* value, size_t rank) {
+  ConstantValueMap::SetRank(value->debugName(), rank);
+  c10::optional<size_t> rank_opt = rank;
+  auto sym_shape = ::c10::SymbolicShape(rank_opt);
+  value->setType(
+      value->type()->cast<TensorType>()->withSymbolicShapes(sym_shape));
+}
+
+void UpdateShape(Value* value, const std::vector<int64_t>& shape_size) {
+  if (shape_size.empty()) {
+    UpdateRank(value, 0);
+    return;
+  }
+  auto vary_shape_size = c10::VaryingShape<int64_t>(shape_size);
+  ConstantValueMap::SetShape(value->debugName(), vary_shape_size);
+  value->setType(value->type()->cast<TensorType>()->withSymbolicShapes(
+      ::c10::SymbolicShape(shape_size)));
+}
+
+// As an addition to onnx shape inference, this function leverages constant
+// folding and a per-Op based process to update rank/shape for the graph, also
+// it update ConstantValueMap accordingly.
+void ComputeConstant(Node* n, int opset_version) {
+  if (n->kind() == ::c10::onnx::Constant) {
+    if (n->kindOf(attr::value) == AttributeKind::t) {
+      at::Tensor const_val = n->t(attr::value);
+      at::Tensor const_val_copy =
+          at::empty(const_val.sizes(), const_val.options());
+      const_val_copy.copy_(const_val);
+      ConstantValueMap::SetValue(n->output()->debugName(), const_val_copy);
+    }
+    return;
+  }
+  auto only_rank_available = false;
+  size_t rank = 0;
+
+  // Constant folding.
+  auto const_fold_val = ComputeConstantFolding(n, opset_version);
+  if (const_fold_val.has_value()) {
+    at::Tensor const_fold_val_copy = at::empty(
+        const_fold_val.value().sizes(), const_fold_val.value().options());
+    const_fold_val_copy.copy_(const_fold_val.value());
+    ConstantValueMap::SetValue(n->output()->debugName(), const_fold_val_copy);
+    UpdateShape(n->output(), const_fold_val_copy.sizes().vec());
```

Comment:
should it return?

Reply:
Added a return to test (Originally I am concerned on `runTorchBackendForOnnx`, let's try).